akka-stream/src/main/scala/akka/stream/scaladsl/Sink.scala

```diff
@@ -732,4 +734,41 @@ object Sink {
   def lazyFutureSink[T, M](create: () => Future[Sink[T, M]]): Sink[T, Future[M]] =
     Sink.fromGraph(new LazySink(_ => create()))
 
+  /**
+   * Handle each stream element with the help of a resource.
+   *
+   * The resource creation function is invoked once when the stream is materialized and the returned resource is passed to
+   * the function `f` for handling each input element. A `null` resource is not allowed and will fail the stream.
+   *
+   * The `close` function is called only once when the upstream finishes or fails. You can do some clean-up here.
+   *
+   * Adheres to the [[ActorAttributes.SupervisionStrategy]] attribute.
+   *
+   * You can configure the default dispatcher for the returning Sink by changing the `akka.stream.materializer.blocking-io-dispatcher` or
+   * set it by using [[ActorAttributes]].
+   *
+   * '''Backpressures when''' when the previous fold function invocation has not yet completed
+   *
+   * '''Completes when''' upstream completes
+   *
+   * '''Cancels when''' never
+   *
+   * @tparam R the type of the resource
+   * @tparam T the type of the output elements
+   * @param create function that creates the resource
+   * @param f function that handles the upstream element with resource.
+   * @param close function that closes the resource
+   */
+  def foldResource[R, T](create: () => R)(f: (R, T) => Unit, close: R => Unit): Sink[T, Future[Done]] =
```

Comment:
comparing to `fold`:
```
def fold[U, T](zero: U)(f: (U, T) => U): Sink[T, Future[U]]
```

why is foldResource signature so different, both with regards to the zero and the materialized value?

Reply:
also, @johanandren had a suggestion to name it `foldWithResource`. What was the conclusion there?
https://github.com/akka/akka/pull/31573#issuecomment-1247727886

Reply:
Yes, it was `foldWithResource` but IIRC @johanandren suggest the `foldResource` somewhere.

Reply:
@patriknw 
`why is foldResource signature so different, both with regards to the zero and the materialized value?` So what's your suggestion?

Reply:
If the purpose is to have something that is the same as `fold`, but with a managed resource it could be
```
def fold[U, T](create: () => R, zero: U)(f: (R, U, T) => U): Sink[T, Future[U]]
```

but I'm not convinced we should add this withResource everywhere. When do we stop? We already have the basic building blocks for it and then the user can combine them.

Reply:
More complicated, but more powerful @johanandren ping, so the U will be something like `bytes written`?
I think the origin issue is to an alignment with `unfoldResource`.

Reply:
I think the idea was to have something like `Flow.mapWithResource` and `Source.unfoldResource` on the sink, not folding a value with the help of a resource but folding the elements "into" a resource, so more like `Sink.foreach` but with a lifecycle.

It can maybe just as well be done with `mapWithResource()` composed with `Sink.last` or `Sink.fold` for the fold value case and `Sink.ignore` for ignoring the output.

Reply:
Then `foldResource` naming is confusing. The API will difficult to learn with too many variants with similar naming. I suggest we close this. Even if we find a good name I don't think we should add too many variants when it's just about convenience that can be achieved with composing two (or  a few) operators.

Reply:
For me, I think this one is fine, if people want the `U`, and that maybe compose with `mapWithResource` and `fold`. The name can be `toResource` as its not folding with a zero:U. 

I would argue that, even developer/users can do many thing with composing or custom graph stage , I don't think many of us can do it right at the first time or very quickly.

I think that what ZIO is currently doing and some people complains about Akka currently not handling resource very well.